src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/prisonperson/profiledetails/ProfileDetailsService.kt

```diff
@@ -54,24 +56,71 @@ class ProfileDetailsService(
             },
         ).takeIf { bookingResponse -> bookingResponse.profileDetails.isNotEmpty() }
       }
-      .filterNotNull()
       .let { PrisonerProfileDetailsResponse(offenderNo = offenderNo, bookings = it) }
   }
 
   fun upsertProfileDetails(offenderNo: String, request: UpsertProfileDetailsRequest): UpsertProfileDetailsResponse {
-    // TODO SDIT-2023 Rewrite this once all tests written. Specifically make it more readable and lose all the !!
-    var created = true
-    val profileType = profileTypeRepository.findByIdOrNull(request.profileType)!!
-    val profileCode = profileCodeRepository.findByIdOrNull(ProfileCodeId(request.profileType, request.profileCode!!))!!
-    val booking = bookingRepository.findLatestByOffenderNomsId(offenderNo)
-    val profile = booking!!.profiles.firstOrNull()
-      ?: OffenderProfile(OffenderProfileId(booking, 1), LocalDate.now())
-    profile.profileDetails.find { it.id.profileType.type == request.profileType }
-      ?.apply { this.profileCodeId = profileCode.id.code }
-      ?.also { created = false }
-      ?: OffenderProfileDetail(OffenderProfileDetailId(booking, profile.id.sequence, profileType), 1L, profile, profileCode)
-        .also { profile.profileDetails += it }
-    offenderProfileRepository.save(profile)
+    validateProfileType(request.profileType)
+    validateProfileCode(request.profileType, request.profileCode)
+
+    val booking = findLatestBookingOrThrow(offenderNo)
+    val (profileDetails, created) = booking.getOrCreateProfile().getOrCreateProfileDetails(request)
+    profileDetails.profileCodeId = request.profileCode
+
     return UpsertProfileDetailsResponse(created, booking.bookingId)
+      .also {
+        telemetryClient.trackEvent(
+          """profile-details-physical-attributes-${if (created) "created" else "updated"}""",
+          mapOf(
+            "offenderNo" to offenderNo,
+            "bookingId" to booking.bookingId.toString(),
+            "profileType" to request.profileType,
+          ),
+          null,
+        )
+      }
+  }
```

Comment:
The main change was to flesh out all lthe tests for this method and refactor it.